common/etc/nginx/include/s3gateway.js

```diff
@@ -643,10 +690,120 @@ function _require_env_var(envVarName) {
     }
 }
 
+/**
+ * Offset to the expiration of credentials, when they should be considered expired and refreshed. The maximum
+ * time here can be 5 minutes, the IMDS and ECS credentials endpoint will make sure that each returned set of credentials
+ * is valid for at least another 5 minutes.
+ *
+ * To make sure we always refresh the credentials instead of retrieving the same again, keep credentials until 4:30 minutes
+ * before they really expire.
+ *
+ * @type {number}
+ */
+var maxValidityOffsetMs = 4.5 * 60 * 100;
+
+async function fetchCredentials(r) {
+    var current = readCredentials();
+    if (current) {
+        var exp = new Date(current.expiration).getTime() - maxValidityOffsetMs;
+        if (now.getTime() < exp) {
+            r.return(200);
+            return;
+        }
+    }
+    _debug_log(r, 'Cached credentials are expired or not present, requesting new ones');
+
+    var credentials;
+    if (process.env['S3_ACCESS_KEY_ID']) {
+        credentials = {
+            accessKeyId: process.env['S3_ACCESS_KEY_ID'],
+            secretAccessKey: process.env['S3_SECRET_KEY'],
+            sessionToken: null,
+        };
+    } else if (process.env['AWS_CONTAINER_CREDENTIALS_RELATIVE_URI']) {
```

Comment:
The simple existence test here is not sufficient as it fails to match in unit tests. Perhaps we should do something like:
```javascript
} else if ('AWS_CONTAINER_CREDENTIALS_RELATIVE_URI' in process.env) {
```

Reply:
Hmm, you sure? I ran the tests once (it is a bit difficult for me to run the tests except in the CI tbh :(), and they worked. During the patches, I changed, however, the env variable back to the one injected by AWS (AWS_CONTAINER_CREDENTIALS_RELATIVE_URI), the unit tests still sets the non-default one (AWS_CONTAINER_CREDENTIALS_ABSOLUTE_URI), I'll change that and then this should work (at least on my newbie understanding, you might still be right 🙈)

Reply:
Sadly, I'm not *that* sure. I did some experiments in njs with trying to add in new values to `process.env` and found that I couldn't do it. How did you run the unit tests? I'm running them by executing:
```
/usr/bin/njs -t module -p common/etc/nginx test/unit/s3gateway_test.js
```

If you run `test.sh` it will also execute the unit tests in addition to the integration tests.

Reply:
I did another simpler test:

```javascript
process.env['FOO'] = 'bar';
console.log(process.env['FOO']);
```

```
$ njs process_env.js 
bar
```

In this case it looks like it is successfully modifying the environment variables. I'm going to dive in a bit deeper.

Reply:
I used test.sh. Let me try to run them again (in both ways). However, I use a windows environment (together with cygwin) 🤦 which made things a bit harder iirc.

Reply:
I just run the tests (after I fixed some more issues with mocking a Response promise 🙈) and they run fine now (the unit tests, in a docker container) :)